server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -87,6 +86,64 @@ public Builder add(Mapper.Builder builder) {
             return this;
         }
 
+        public Builder addMappers(Map<String, Mapper> mappers) {
+            mappers.forEach((name, mapper) -> mappersBuilders.add(new Mapper.Builder(name) {
+                @Override
+                public Mapper build(MapperBuilderContext context) {
+                    return mapper;
+                }
+            }));
+            return this;
+        }
+
+        /**
+         * Adds a dynamically created Mapper to this builder.
+         *
+         * @param name      the name of the Mapper, including object prefixes
+         * @param prefix    the object prefix of this mapper
+         * @param mapper    the mapper to add
+         * @param context   the DocumentParserContext in which the mapper has been built
+         */
+        public void addDynamic(String name, String prefix, Mapper mapper, DocumentParserContext context) {
+            // If the mapper to add has no dots and is therefore
+            // a leaf mapper, we just add it here
+            if (name.contains(".") == false) {
+                mappersBuilders.add(new Mapper.Builder(name) {
+                    @Override
+                    public Mapper build(MapperBuilderContext context) {
+                        return mapper;
+                    }
+                });
+            }
+            // otherwise we strip off the first object path of the mapper name, load or create
+            // the relevant object mapper, and then recurse down into it, passing the remainder
+            // of the mapper name. So for a mapper 'foo.bar.baz', we locate 'foo' and then
+            // call addDynamic on it with the name 'bar.baz'.
+            else {
+                int firstDotIndex = name.indexOf(".");
+                String childName = name.substring(0, firstDotIndex);
+                String fullChildName = prefix == null ? childName : prefix + "." + childName;
+                ObjectMapper.Builder childBuilder = findChild(childName, fullChildName, context);
+                childBuilder.addDynamic(name.substring(firstDotIndex + 1), fullChildName, mapper, context);
+                mappersBuilders.add(childBuilder);
```

Comment:
I'm probably confused by the recursion here, but would it make sense to add the new childBuilder before the recursive step? Or maybe it doesn't matter, I now see that "findChild" only looks up previous dynamic updates in 'context.getObjectMapper(fullChildName)' and that wouldn't be affected by the `mappersBuilders` here? Just a bit confused, maybe you can explain.

Reply:
Yes, `findChild` doesn't actually consult the existing `mappersBuilders` at all.  In fact, I think the final branch of `findChild`, where we just return a new builder, won't ever be called now, because of #79922 - a dynamically built mapper will always have a parent that is either in the existing mappings, or has also been dynamically added.  Maybe worth keeping it as a safety fall-back though.

In terms of adding the childBuilder to mappersBuilders before/after the recursion, it doesn't actually matter.   The recursive step knows nothing about its parent (apart from using the passed-in prefix to build a full child name for lookup purposes), and the childBuilder itself is mutable so we can add a new mapper to it either before or after it has been added to `mappersBuilders`, it's the same object with the same result.

Reply:
I think the thing the feels weird to me is that we're merging stuff from the tree of updates in the context and the tree under the object itself. It's, like, how you have to do it, I think. I left a note about the method itself below. Maybe I've found something? If I have maybe it'd give us something to pry on to simplify this.

Reply:
It's not quite under the object itself, it's under a new, empty, objectbuilder that has been generated by the main object.  I've tried to explain the recursion more fully in a comment below.

Reply:
Interestingly, if I remove the third branch and throw an exception, it causes failures in tests that are checking how fields starting or ending with a dot behave.  I'll see if I can pull field name validation out into a separate PR.